src/utils.jl

```diff
@@ -246,66 +101,14 @@ form is disabled and the 2-argument signature is ensured to be matched.
 function isinplace(f, inplace_param_number, fname = "f", iip_preferred = true;
         has_two_dispatches = true, isoptimization = false,
         outofplace_param_number = inplace_param_number - 1)
-    nargs = numargs(f)
-    iip_dispatch = any(x -> x == inplace_param_number, nargs)
-    oop_dispatch = any(x -> x == outofplace_param_number, nargs)
-
-    if length(nargs) == 0
-        throw(NoMethodsError(fname))
-    end
-
-    if !iip_dispatch && !oop_dispatch && !isoptimization
-        if all(>(inplace_param_number), nargs)
-            throw(TooManyArgumentsError(fname, f))
-        elseif all(<(outofplace_param_number), nargs) && has_two_dispatches
-            # Possible extra safety?
-            # Find if there's a `f(args...)` dispatch
-            # If so, no error
-            _parameters = if methods(f).ms[1].sig isa UnionAll
-                Base.unwrap_unionall(methods(f).ms[1].sig).parameters
-            else
-                methods(f).ms[1].sig.parameters
-            end
-
-            for i in 1:length(nargs)
-                if nargs[i] < inplace_param_number &&
-                   any(isequal(Vararg{Any}), _parameters)
-                    # If varargs, assume iip
-                    return iip_preferred
-                end
-            end
-
-            # No varargs detected, error that there are dispatches but not the right ones
-
-            throw(TooFewArgumentsError(fname, f, isoptimization))
-        else
-            throw(FunctionArgumentsError(fname, f))
-        end
-    elseif oop_dispatch && !iip_dispatch && !has_two_dispatches
-
-        # Possible extra safety?
-        # Find if there's a `f(args...)` dispatch
-        # If so, no error
-        for i in 1:length(nargs)
-            if nargs[i] < inplace_param_number &&
-               any(isequal(Vararg{Any}), methods(f).ms[1].sig.parameters)
-                # If varargs, assume iip
-                return iip_preferred
-            end
-        end
-
-        throw(TooFewArgumentsError(fname, f, isoptimization))
+    if iip_preferred
+        hasmethod(f, ntuple(_->Any, inplace_param_number)) && return true
+        hasmethod(f, ntuple(_->Any, outofplace_param_number)) && return false
     else
-        if iip_preferred
-            # Equivalent to, if iip_dispatch exists, treat as iip
-            # Otherwise, it's oop
-            iip_dispatch
-        else
-            # Equivalent to, if oop_dispatch exists, treat as oop
-            # Otherwise, it's iip
-            !oop_dispatch
-        end
+        hasmethod(f, ntuple(_->Any, outofplace_param_number)) && return false
+        hasmethod(f, ntuple(_->Any, inplace_param_number)) && return true
     end
+    throw(FunctionArgumentsError(fname, f))
```

Comment:
This error message loses a lot of information. Can we in the error path do a method check and throw the more informative error message?

Reply:
IMO the better method is to just put the expected `nargs` in the exception and then we can make the exception printing in charge of that logic. That way none of the complicated bits are in the main function.